packages/remix-react/__tests__/hook-types-test.tsx

```diff
@@ -100,4 +100,14 @@ describe("type serializer", () => {
     type response = UseDataFunctionReturn<Loader>;
     isEqual<response, { arg: string }>(true);
   });
+
+  it("makes keys optional if the value is undefined", () => {
+    type AppData = {
+      arg1: string;
+      arg2: number | undefined;
+      arg3: undefined;
```

Comment:
The alternative would be to have users who want optional properties to use Typescript optionals:

```suggestion
      arg2?: number;
      arg3?: undefined; // probably would just delete this param altogether
```

My concern is that Typescript _does_ distinguish between optional properties and properties that are unions with `undefined`, so if any users want to rely on standard Typescript behavior, we'd be taking that away from them.

Is this feature to help out more novice users who aren't aware of Typescript optionals? Or do you also see upside for users aware of Typescript optionals?

Reply:
The `AppData` type in this test isn't something the user would write themselves. It represents the use case of passing something possibly undefined to the browser. Here's an example:

```ts
declare function getUser(): User | undefined

export function loader() {
  return json({ user: getUser() })
}

export default function Page() {
	const data = useLoaderData()
}
```

Before this PR, the type of `data` would become `{ user: User }`, which is incorrect. It should be `{ user?: User }`

The `arg3: undefined` is only here for completeness. If something is undefined, it can't be JSON-serialized, and doesn't show up in the object at all.

Reply:
I meant the difference between `{a?: T}` and `{a: T | undefined}` (note that `a` is not optional in the second type).

But yes, the `getUser` example is compelling. Especially since `undefined` is invalid JSON and gets dropped in `JSON.stringify`. That convinces me!